### ISSUE TITLE
Native C# WPF: Method could not be found: SetValue

### ISSUE THREAD

**Issue by JohnLouderback** [trueinteractions/tint2#127]
> There is a group of methods on DependencyObject (base of most WPF classes) which allow the 
> common access to all dependency properties. They are
>
> * SetValue
> * GetValue
> * ClearValue

When I try to use the method "SetValue" on the native object of a UI component, I get the error: ``Method could not be found: SetValue``

**Comment by trevorlinton** [comment]
Hi John,

The set value method is available on the native objects, but there may be a few things to take a look at/consider when using it:

Its an overloaded method, if the CLR/Javascript cannot easily convert the types pushed into SetValue it will throw an exception. If you find yourself passing in a JS value (not a derived C# class/structure) you might want to try using an explicit type.

Even though it exists on the dependecy/propertyinfo class the definition there is actually just a place holder for docs/compilers to validate on, the actual resolution of the property is done at runtime via refleciton.

**SO**

You can post an example of what you're trying to do and perhaps I can help shed some light on it.  We use SetValue quite a bit all over Tint, but perhaps theres a corner case where its not working.

See https://github.com/trueinteractions/tint2/blob/2134842dd3c16335b200549b904937120443833e/modules/Window/Window_win.js#L28 for an example of how its being used.

I'll leave this open and let you close it out when you feel its been adequately addressed.

**Comment by JohnLouderback** [comment]
Hi Trevor,

Thanks for the information. That gives me a lot more insight into this. So, I'm guessing the case this that there is no method signature that matches the values I'm passing in. The code I'm attempting to use is:

```javascript
Chromium = require('chromium');
var webview = new Chromium();
webview.nativeView.SetValue($.System.Windows.Shell.WindowChrome.IsHitTestVisibleInChrome, true);
```

I'm inclined to think that maybe the JS boolean isn't getting cast to the explicit C# type ``System.Boolean``.

**Comment by trevorlinton** [comment]
@JohnLouderback 

Hi John,

I looked for a bit and it seems you're correct, the reflection fails as bool converts to a native type while a dependency property expects an Object (or System.Boolean).  You can use System.Boolean and use the Parse method to parse a string to a true to get a Boolean object but using dependency properties generally always have a static "setter" and "getter" on the property type.  The IsHitTestVisibleInChrome is available here as a static method: https://msdn.microsoft.com/en-us/library/system.windows.shell.windowchrome.setishittestvisibleinchrome(v=vs.110).aspx

I gave it a try and didnt have any issues:
```$.System.Windows.Shell.WindowChrome.SetIsHitTestVisibleInChrome(webview.nativeView, true);```

Notice how it takes a bool and not a System.Boolean.  This is one of the design flaws of dependency properties which I suspect .NET resolves through the compiler.  We can check for bool and try both bool and System.Boolean during the reflection to hopefully mitigate these issues, but we'd have to do it for every method call... which is a lot when you consider you may have over 100,000 method calls in a programs execution, so we'd like to try and keep that somewhat tight.  I'll see if there's any easier work arounds to get this to function properly, but in the meantime you can use the static setters/getters approach to directly set boolean attached properties.

Feel free to close this issue out once you feel its been resolved.

Thanks,
Trevor

**Comment by JohnLouderback** [comment]
Hi Trevor, that makes sense. The above snippet works great too. This is resolved as far as I'm concerned. It's a shame that there doesn't appear to be a way to know whether a JS boolean is required or a C# boolean. I don't know what the reasonable limitations of the reflection bridge API are, but maybe there's a way to check if other methods with that name but a different signature exist and display them and their signatures if the method isn't found. Just an idea. In any case, just being aware of this makes a big difference for me

**Comment by trevorlinton** [comment]
@JohnLouderback I agree, there might be a way when we load the bridge to inspect if the method is expecting an object and interject a javascript bool -> System.Boolean interop. This would avoid the penalty of checking it at runtime and honestly the easier we make the interface/bridge the better adoption it'll have.  I'm going to leave this open until we have this properly resolved.  If you have any other feedback feel free to post it here.
